### PR TITLE
Feature/change collectpads to non waiting

### PR DESCRIPTION
**Pull request by rrcarlosrodriguez**

### CODE REVIEW

File: gst-libs/gst/r2inference/gstvideoinference.c [hunk 1]
```diff
@@ -861,204 +881,340 @@ video_inference_transform_meta (GstBuffer * buffer_model,
...
-  ret =
-      gst_video_inference_pop_buffer (self, pads,
-      (GstCollectData *) priv->sink_bypass_data, &buffer_bypass);
-  if (GST_FLOW_OK != ret) {
-    goto model_free;
+  GST_LOG_OBJECT (self, "Processing bypass buffer");
+
+  /* Check if model pad was requested, if not, forward buffer */
+  if (NULL == priv->sink_model) {
+    GST_INFO_OBJECT (self,
```
Comment (michaelgruner) [inline gst-libs/gst/r2inference/gstvideoinference.c]: Way too high for every buffer. Demote to log

File: gst-libs/gst/r2inference/gstvideoinference.c [hunk 2]
```diff
@@ -861,204 +881,340 @@ video_inference_transform_meta (GstBuffer * buffer_model,
...
-      goto bypass_free;
+  bypass_buffer = gst_buffer_make_writable (buffer);
+  current_meta = gst_buffer_get_meta (bypass_buffer,
+      gst_inference_meta_api_get_type ());
+  if (current_meta) {
+    /* Check if at least one node is enabled to be processed, if not, just forward buffer */
+    GstInferenceMeta *imeta = (GstInferenceMeta *) current_meta;
+    gboolean enabled = FALSE;
+
+    g_node_traverse (imeta->prediction->predictions, G_LEVEL_ORDER,
```
Comment (michaelgruner) [inline gst-libs/gst/r2inference/gstvideoinference.c]: Make this part of the prediction api

File: gst-libs/gst/r2inference/gstvideoinference.c [hunk 3]
```diff
@@ -861,204 +881,340 @@ video_inference_transform_meta (GstBuffer * buffer_model,
...
+          "There is no predictions enabled, bypassing processing...");
+      goto forward_buffer;
     }
+  }
+
+  /* Queue this new buffer at the head and dequeue the older one */
+  GST_LOG_OBJECT (self, "Queue bypass buffer and get older one");
+  g_mutex_lock (&priv->mtx_bypass_queue);
+  g_queue_push_head (priv->bypass_queue, (gpointer) bypass_buffer);
+  bypass_buffer = (GstBuffer *) g_queue_pop_tail (priv->bypass_queue);
```
Comment (michaelgruner) [inline gst-libs/gst/r2inference/gstvideoinference.c]: `GST_BUFFER_CAST`

File: gst-libs/gst/r2inference/gstvideoinference.c [hunk 4]
```diff
@@ -861,204 +881,340 @@ video_inference_transform_meta (GstBuffer * buffer_model,
...
+  g_mutex_unlock (&priv->mtx_bypass_queue);
+
+  while (!model_empty) {
+    /* Dequeue oldest model buffer from tail */
+    g_mutex_lock (&priv->mtx_model_queue);
+    GST_LOG_OBJECT (self, "Dequeue model buffer");
+    model_buffer = (GstBuffer *) g_queue_pop_tail (priv->model_queue);
```
Comment (michaelgruner) [inline gst-libs/gst/r2inference/gstvideoinference.c]: `GST_BUFFER_CAST`

File: gst-libs/gst/r2inference/gstvideoinference.c [hunk 5]
```diff
@@ -861,204 +881,340 @@ video_inference_transform_meta (GstBuffer * buffer_model,
...
+          g_queue_push_tail (priv->model_queue, (gpointer) model_buffer);
+          g_mutex_unlock (&priv->mtx_model_queue);
+          goto forward_buffer;
+        }
+      }
+
+      /* Transfer meta from model to bypass */
+      GST_LOG_OBJECT (self, "Transfering meta from model to bypass");
+
+      meta_bypass[0] =
```
Comment (michaelgruner) [inline gst-libs/gst/r2inference/gstvideoinference.c]: Why is this being done even if model_buffer is NULL?

File: gst-libs/gst/r2inference/gstvideoinference.c [hunk 6]
```diff
@@ -806,6 +818,23 @@ video_inference_prepare_postprocess (const GstMetaInfo * meta_info,
     return TRUE;
   }
 
+  if (out_meta) {
```
Comment (michaelgruner) [inline gst-libs/gst/r2inference/gstvideoinference.c]: Now, out_meta is always passed in.

File: gst-libs/gst/r2inference/gstvideoinference.c [hunk 7]
```diff
@@ -861,204 +881,340 @@ video_inference_transform_meta (GstBuffer * buffer_model,
...
 
+  /* Assign already created inferencemeta, no need to create a new one */
+  meta_model[1] = current_meta;
+
+  /* Prepare postprocess */
+  info_model = &(pad->info);
   if (!video_inference_prepare_postprocess (klass->inference_meta_info,
-          buffer_bypass, info_bypass, &frame_bypass)) {
-    return FALSE;
+          buffer_model, info_model, &frame_model, meta_model)) {
```
Comment (michaelgruner) [inline gst-libs/gst/r2inference/gstvideoinference.c]: Frame is not being used

File: gst-libs/gst/r2inference/gstvideoinference.c [hunk 8]
```diff
@@ -861,204 +881,340 @@ video_inference_transform_meta (GstBuffer * buffer_model,
...
 
-    /* Have the subclass analyze the prediction and generate model and bypass metas */
-    if (!gst_video_inference_postprocess (self, klass, prediction_data,
-            prediction_size, buffer_model, priv->sink_model_data, buffer_bypass,
-            priv->sink_bypass_data)) {
-      ret = GST_FLOW_ERROR;
-      goto bypass_free;
+      if (GST_FLOW_OK != ret) {
+        /* Free if forward failed */
+        video_inference_buffer_unref (model_buffer);
```
Comment (michaelgruner) [inline gst-libs/gst/r2inference/gstvideoinference.c]: Double check if these are still needed.

File: gst-libs/gst/r2inference/gstvideoinference.c (hunk 5 above)
Comment (rrcarlosrodriguez) [reply gst-libs/gst/r2inference/gstvideoinference.c]: No, if model_buffer is NULL, we stop the while loop, we don't pass through this code

File: gst-libs/gst/r2inference/gstvideoinference.c [hunk 9]
```diff
@@ -861,204 +860,301 @@ video_inference_transform_meta (GstBuffer * buffer_model,
...
-gst_video_inference_postprocess (GstVideoInference * self,
-    GstVideoInferenceClass * klass, const gpointer prediction_data,
-    gsize prediction_size, GstBuffer * buffer_model,
-    GstVideoInferencePad * pad_model, GstBuffer * buffer_bypass,
-    GstVideoInferencePad * pad_bypass)
+static GstFlowReturn
+gst_video_inference_process_model (GstVideoInference * self, GstBuffer * buffer,
+    GstVideoInferencePad * pad)
 {
+  GstVideoInferencePrivate *priv = GST_VIDEO_INFERENCE_PRIVATE (self);
```
Comment (tvlenin) [inline gst-libs/gst/r2inference/gstvideoinference.c]: Use self before the pointer guard.

File: gst-libs/gst/r2inference/gstvideoinference.c [hunk 10]
```diff
@@ -861,204 +860,301 @@ video_inference_transform_meta (GstBuffer * buffer_model,
...
+  GstVideoFrame frame_bypass;
+  GstVideoInfo *info_model = NULL;
+  GstVideoInfo *info_bypass = NULL;
+  GstMapFlags flags;
 
-  if (NULL == data) {
-    return GST_FLOW_OK;
-  }
+  g_return_if_fail (model_buffer);
+  g_return_if_fail (meta_model);
```
Comment (tvlenin) [inline gst-libs/gst/r2inference/gstvideoinference.c]: Pointer guards bypass_buffer, self, meta_bypass.

**Pull request merged**